src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs

```diff
@@ -42,97 +42,80 @@ private enum GlobalNotificationState
+                    // make sure connection is done right
+                    var host = await client._rpc.InvokeWithCancellationAsync<string>(
+                        nameof(IRemoteHostService.Connect), new object[] { clientId, uiCultureLCID, cultureLCID, TelemetryService.DefaultSession.SerializeSettings() }, cancellationToken).ConfigureAwait(false);
 
-                var uiCultureLCID = CultureInfo.CurrentUICulture.LCID;
-                var cultureLCID = CultureInfo.CurrentCulture.LCID;
+                    client.Started();
 
-                // make sure connection is done right
-                var host = await client._rpc.InvokeWithCancellationAsync<string>(
-                    nameof(IRemoteHostService.Connect), new object[] { current, uiCultureLCID, cultureLCID, TelemetryService.DefaultSession.SerializeSettings() }, cancellationToken).ConfigureAwait(false);
+                    return client;
+                }
+                catch (ConnectionLostException ex)
+                {
+                    RemoteHostCrashInfoBar.ShowInfoBar(workspace, ex);
 
-                return client;
-            }
-            catch (ConnectionLostException ex)
-            {
-                RemoteHostCrashInfoBar.ShowInfoBar(workspace, ex);
+                    Shutdown(ex);
 
-                Shutdown(client, ex, cancellationToken);
+                    // dont crash VS because OOP is failed to start. we will show info bar telling users to restart
+                    // but never physically crash VS.
+                    return null;
+                }
+                catch (SoftCrashException ex)
+                {
+                    Shutdown(ex);
 
-                // dont crash VS because OOP is failed to start. we will show info bar telling users to restart
-                // but never physically crash VS.
-                throw new SoftCrashException("Connection Lost", ex, cancellationToken);
-            }
-            catch (Exception ex)
-            {
-                Shutdown(client, ex, cancellationToken);
-                throw;
-            }
+                    // at this point, we should have shown info bar (RemoteHostCrashInfoBar.ShowInfoBar) to users
+                    // returning null here will disable OOP for this VS session. 
+                    // * Note * this is not trying to recover the exception. but giving users to time
+                    // to clean up before restart VS
+                    return null;
+                }
+                catch (Exception ex)
+                {
+                    Shutdown(ex);
+                    throw;
+                }
 
-            static void Shutdown(ServiceHubRemoteHostClient? client, Exception ex, CancellationToken cancellationToken)
-            {
-                // make sure we shutdown client if initializing client has failed.
-                client?.Shutdown();
+                void Shutdown(Exception ex)
```

Comment:
please make static.  it makes it clear what local state is being affected by the local function.  i actually fonud the above code confusing because i couldn't understand how shutdown could work properly.

Reply:
I think this is better.

Reply:
Note: i legit could not understand the code in the method.  It was really strange and actually made no sense.  There is zero way to tell about these sorts of captures, so all behavior is "spooky at a distance".  When we started adopting local functions, being static as much as possible has been far preferred because it makes the code extremely clear and calls at *at the callsite* (no pun intended): these variables will be affected here.  That's a big plus for readability and maintenance.